service/storage_service.cc

```diff
@@ -6119,6 +6155,180 @@ future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
+            default:
+                result.result = join_node_request_result::accepted {};
+                co_return result;
+            }
+        }
+
+        if (params.replaced_id.has_value()) {
+            if (!_topology_state_machine._topology.normal_nodes.contains(*params.replaced_id)) {
+                result.result = join_node_request_result::rejected {
+                    .reason = ::format("Cannot replace node {} because it is not in the 'normal' state", *params.replaced_id),
+                };
+                co_return result;
+            }
+        }
+
+        auto mutation = build_mutation_from_join_params(params, *guard);
+
+        topology_change change{{std::move(mutation)}};
+        group0_command g0_cmd = _group0->client().prepare_command(std::move(change), *guard,
+                format("raft topology: placing join request for {}", params.host_id));
+        try {
+            co_await _group0->client().add_entry(std::move(g0_cmd), std::move(*guard), &_abort_source);
+            break;
+        } catch (group0_concurrent_modification&) {
+            slogger.info("raft topology: join_node_request: concurrent operation is detected, retrying.");
+        }
+    }
+
+    slogger.info("raft topology: placed join request for {}", params.host_id);
+
+    // Success
+    result.result = join_node_request_result::ok {};
+    co_return result;
+}
+
+future<join_node_response_result> storage_service::join_node_response_handler(join_node_response_params params) {
+    assert(this_shard_id() == 0);
+
+    // Serialize handling the responses
+    auto lock = co_await get_units(_join_node_response_handler_mutex, 1);
+
+    // Wait until the join_node_request is sent.
+    co_await _join_node_request_done.get_shared_future(_abort_source);
+
+    if (_join_node_result.available()) {
+        // We already handled this RPC. Return immediately for idempotence.
+        join_node_response_result result;
+        co_return result;
+    }
+
+    co_return co_await std::visit(overloaded_functor {
+        [&] (const join_node_response_params::accepted& acc) -> future<join_node_response_result> {
+            // Do a read barrier to read/initialize the topology state
+            auto& raft_server = _group0->group0_server();
+            co_await raft_server.read_barrier(&_abort_source);
+
+            // Calculate nodes to ignore
+            // TODO: ignore_dead_nodes setting for bootstrap
+            std::unordered_set<raft::server_id> ignored_ids;
+            auto my_request_it =
+                    _topology_state_machine._topology.req_param.find(_group0->load_my_id());
+            if (my_request_it != _topology_state_machine._topology.req_param.end()) {
+                if (auto* replace = std::get_if<service::replace_param>(&my_request_it->second)) {
+                    ignored_ids = replace->ignored_ids;
+                    ignored_ids.insert(replace->replaced_id);
+                }
+            }
```

Comment:
Can ignored nodes be taken from the config params instead of through the group0?

Reply:
I originally wanted to do it like this, but we need `replacement_info` which is calculated in `join_token_ring` and I didn't see a way to pass it to the rpc handler in a non-awkward way (add a field to storage_service? register the rpc verb in join_token_ring and capture replacement_info with a lambda?).

Reply:
Yes, good point. Fine to get them from the state machine indeed.